File: pytools/examples/python/server/network/TcpClient.py
#!/usr/bin/env python
# coding:utf-8
from Packet import Packet
import packet_util

from protocol import msg, pb


class TcpClient(object):
	__socket 			= None

	__head_len			= 2
	__buffers			= ''
	__buffers_len 		= 0
	__buffers_tmp_len	= 512

	def __init__(self, socket):
		self.__socket = socket

	def run(self):
		if not self.__socket:
			raise('socket is None')

		while True:
			buffers_tmp = self.__socket.recv(self.__buffers_tmp_len)
			self.__buffers += buffers_tmp
			self.__buffers_len += len(buffers_tmp)

			while True:
				if self.__buffers_len > self.__head_len:
					head_buff 	= self.__buffers[:2]
					package_len	= packet_util.read_u16(head_buff)
					if self.__buffers_len >= self.__head_len + package_len:
						packet_id_buff 	= self.__buffers[2:4]
						packet_buff 	= self.__buffers[4:package_len + self.__head_len + 2]
						packet_id 		= packet_util.read_u16(packet_id_buff)

						body_len		= self.__head_len + package_len

						self.__buffers 	= self.__buffers[body_len:]
						self.__buffers_len -= body_len
						packet 			= Packet(packet_buff)
						self.dispath(packet_id, packet)
					else:
						break
				else:
					break

	def dispath(self, packet_id, packet):
		print
		print('packet_id:', packet_id)
		print
		if packet_id == msg.P_REQ_TEST_X_X:
			reqTestXX = pb.ReqTestXX(packet)
			print('reqTestXX.id_u8:', reqTestXX.id_u8)
			print('reqTestXX.id_u16: ', reqTestXX.id_u16)
			print('reqTestXX.id_u32: ', reqTestXX.id_u32)
			print('reqTestXX.repeat_id_u8: ', reqTestXX.repeat_id_u8)
			print('reqTestXX.optional_id_u8: ', reqTestXX.optional_id_u8)

			ackTestXX = pb.AckTestXX()
			ackTestXX.id_u8 = 255
			ackTestXX.id_u16 = 65535
			ackTestXX.id_u32 = 65535666
			ackTestXX.repeat_id_u8 = [254,255]
			ackTestXX.optional_id_u8 = 255
			self.send(ackTestXX.encode())

		elif packet_id == msg.P_REQ_TEST_SEND:
			reqTestSend = pb.ReqTestSend(packet)
			print('reqTestSend.id_u8:', reqTestSend.id_u8)
			print('reqTestSend.id_u16: ', reqTestSend.role_base)
			print('reqTestSend.id_u32: ', reqTestSend.id_f32)
			print('reqTestSend.id_op_u8: ', reqTestSend.id_op_u8)
			print('reqTestSend.optional_id_u8: ', reqTestSend.op_role_base)

			msgRoleBase = pb.MsgRoleBase()
			msgRoleBase.uid = 10086
			msgRoleBase.uname = 'mirahs'

			ackTestSendOk = pb.AckTestSendOk()
			ackTestSendOk.id_u8 = 255
			ackTestSendOk.role_base = msgRoleBase
			ackTestSendOk.id_f32 = [1.23,1.24]
			ackTestSendOk.id_op_u8 = 255
			ackTestSendOk.op_role_base = msgRoleBase
			self.send(ackTestSendOk.encode())



	def send(self, buff):
		self.__socket.send(buff)

File: pytools/examples/java/server/network/TcpClient.java
package network;


import protocol.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import java.util.List;
import java.util.ArrayList;

import java.net.Socket;
import java.net.UnknownHostException;


public class TcpClient extends Thread
{
	private Socket socket = null;
	private InputStream input = null;
	private OutputStream output = null;


	public TcpClient(Socket socket)
	{
		this.socket = socket;
		try
		{
			this.input = this.socket.getInputStream();
			this.output = this.socket.getOutputStream();
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}


	public void run()
	{
		try
		{
			int header_len = 2;
			int max_buffer_len = 102400;
			int tmp_buffer_len = 1024;
			int buffer_len = 0;
			byte[] buffer = new byte[max_buffer_len];
			byte[] tmp_buffer = new byte[tmp_buffer_len];

			System.out.println("connect success......");

			while (true)
			{
				int readLen = this.input.read(tmp_buffer);
				//System.out.println("readLen: " + readLen);
				System.arraycopy(tmp_buffer, 0, buffer, buffer_len, readLen);
				buffer_len += readLen;
				//System.out.println("buffer_len: " + buffer_len);

				if (buffer_len <= header_len)
				{
					continue;
				}
				while (true)
				{
					int package_len = PacketUtil.readUShort(buffer);
					//System.out.println("package_len: " + package_len);
					if (buffer_len < package_len + header_len)
					{
						break;
					}
					buffer_len -= package_len + header_len;
					byte[] buf = new byte[package_len];
					System.arraycopy(buffer, header_len, buf, 0, package_len);
					//System.arraycopy(buffer, 0, buffer, header_len + package_len, buffer_len);
					System.arraycopy(buffer, header_len + package_len, buffer, 0, buffer_len);
					//System.out.println("buffer_len: " + buffer_len);

					Packet packet = new Packet(buf);
					doMsg(packet);
				}
			}
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}


	private void doMsg(Packet packet)
	{
		short packetIdTmp = packet.readShort();
		int packetId = PacketUtil.readUShort(packetIdTmp);
		System.out.println();
		System.out.println("packetId: " + packetId);

		try
		{
			switch (packetId)
			{
				case Msg.P_REQ_TEST_X_X:
				{
					ReqTestXX reqTestXX = new ReqTestXX(packet);
					System.out.println("reqTestXX.getIdU8(): " + reqTestXX.getIdU8());
					System.out.println("reqTestXX.getIdU16(): " + reqTestXX.getIdU16());
					System.out.println("reqTestXX.getIdU32(): " + reqTestXX.getIdU32());


					AckTestXX ackTestXX = new AckTestXX();
					ackTestXX.setIdU8((byte)255);
					ackTestXX.setIdU16((short)65535);
					ackTestXX.setIdU32(65535666);

					List<Byte> id_u8 = new ArrayList<Byte>();
					id_u8.add((byte)254);
					id_u8.add((byte)255);

					ackTestXX.setRepeatIdU8(id_u8);
					ackTestXX.setOptionalIdU8((byte)255);

					this.output.write(ackTestXX.encode());
					this.output.flush();

					break;
				}
				case Msg.P_REQ_TEST_SEND:
				{
					ReqTestSend reqTestSend = new ReqTestSend(packet);
					System.out.println("reqTestSend.getIdU8(): " + reqTestSend.getIdU8());
					System.out.println("reqTestSend.getRoleBase().getUid(): " + reqTestSend.getRoleBase().getUid());
					System.out.println("reqTestSend.getRoleBase().getUname(): " + reqTestSend.getRoleBase().getUname());
					System.out.println("reqTestSend.getOpRoleBase().getUid(): " + reqTestSend.getOpRoleBase().getUid());
					System.out.println("reqTestSend.getOpRoleBase().getUname(): " + reqTestSend.getOpRoleBase().getUname());


					MsgRoleBase msgRoleBase = new MsgRoleBase();
					msgRoleBase.setUid(12306);
					msgRoleBase.setUname("mirahs");

					List<Float> id_f32 = new ArrayList<Float>();
					id_f32.add(1.33f);
					id_f32.add(1.44f);

					AckTestSendOk ackTestSendOk = new AckTestSendOk();
					ackTestSendOk.setIdU8((byte)255);
					ackTestSendOk.setRoleBase(msgRoleBase);
					ackTestSendOk.setIdF32(id_f32);
					ackTestSendOk.setIdOpU8((byte)12);
					ackTestSendOk.setOpRoleBase(msgRoleBase);

					
					this.output.write(ackTestSendOk.encode());
					this.output.flush();

					break;
				}
			}
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}
	
}
